kmp_cpu.cpp

```diff
@@ -3,30 +3,66 @@
 static void get_fail(const char *pattern, int pattern_length, int *fail) {
     int candidate = 0;
     fail[0] = -1;
+    unsigned char clear_mask = 0x00000011;
     for (int pos = 1; pos < pattern_length; pos++, candidate++) {
-        if (pattern[pos] == pattern[candidate]) {
+        unsigned char com_pos = (pattern[pos / 4] >> (2 * (pos % 4))) | clear_mask;
+        unsigned char com_cand = (pattern[candidate / 4] >> (2 * (candidate % 4))) | clear_mask;
+        if (com_pos == com_cand) {
             fail[pos] = fail[candidate];
         }
         else {
             fail[pos] = candidate;
-            while (candidate >= 0 && pattern[pos] != pattern[candidate]) {
+            while (candidate >= 0 && com_pos != com_cand) {
                 candidate = fail[candidate];
             }
         }
     }
     fail[pattern_length] = candidate;
 }
 
+void char_compress(const char *text, int text_length, char& text_com) {
+    // A: 00; T: 01; C:10; G: 11 
+    unsigned char A_mask = 0x00000000;
+    unsigned char T_mask = 0x00000001;
+    unsigned char C_mask = 0x00000010;
+    unsigned char G_mask = 0x00000011;
+    for (int idx = 0; idx < text_length; idx++) {
+        int i = idx / 4;
+        int j = idx % 4;
+        char text_tmp = text[4 * i + j];
+        if (text_tmp == 'A') {
+            text_com[i] |= (A_mask << (2 * j));
+        }
+        else if (text_tmp == 'T') {
+            text_com[i] |= (T_mask << (2 * j));
+        }
+        else if (text_tmp == 'C') {
+            text_com[i] |= (C_mask << (2 * j));
+        }
+        else if (text_tmp == 'G') {
+            text_com[i] |= (G_mask << (2 * j));
+        }
+    }
+}
+
 int KMP_search(
-    const char *text, int text_length, const char *pattern, int pattern_length,
-    int *output, int max_output_cnt, int *fail
+    const char *text, int16 text_length, const char *pattern, int16 pattern_length,
```

Comment:
The type of 16 bit int is called `int16_t`.

Reply:
changed